src/lib/comms/sol-lwm2m.c

```diff
@@ -4184,6 +4363,384 @@ sol_lwm2m_bootstrap_server_send_finish(struct sol_lwm2m_bootstrap_server *server
+                obj_ctx->obj->id, obj_instance->id, r);
+            return SOL_COAP_RESPONSE_CODE_BAD_REQUEST;
+        }
+
+        obj_instance->should_delete = true;
+        ret = SOL_COAP_RESPONSE_CODE_DELETED;
+    } else {
+        SOL_VECTOR_FOREACH_IDX (&client->objects, obj_ctx, i) {
+            if (!obj_ctx->obj->del) {
+                SOL_WRN("The object %" PRIu16 " does not implement the delete method."
+                    " Skipping this Object.",
+                    obj_ctx->obj->id);
+                continue;
+            }
+
+            SOL_VECTOR_FOREACH_IDX (&obj_ctx->instances, obj_instance, j) {
+                r = obj_ctx->obj->del((void *)obj_instance->data,
+                    (void *)client->user_data, client, obj_instance->id);
+                if (r < 0) {
+                    SOL_WRN("Could not properly delete object id %"
+                        PRIu16 " instance id: %" PRIu16 " reason:%d",
+                        obj_ctx->obj->id, obj_instance->id, r);
+                } else {
+                    ret = SOL_COAP_RESPONSE_CODE_DELETED;
+                }
+            }
+        }
+    }
+
+    return ret;
+}
+
+static uint8_t
+handle_bootstrap_write_resource(struct sol_lwm2m_client *client,
+    struct obj_ctx *obj_ctx, struct obj_instance *obj_instance,
+    struct resource_ctx *resource_ctx, int32_t resource_id,
+    uint16_t content_format, const struct sol_str_slice payload)
+{
+    int r;
+
+    //If bootstrap_write_resource is not NULL then bootstrap_write_resource_tlv
+    //is guaranteed to be valid as well.
+    if (!obj_ctx->obj->write_resource) {
+        SOL_WRN("Object %" PRIu16 " does not support the Bootstrap Write method for Resource",
+            obj_ctx->obj->id);
+        return SOL_COAP_RESPONSE_CODE_BAD_REQUEST;
+    }
+
+    if (!content_format) {
+        SOL_WRN("Content format was not set."
+            " Impossible to write resource");
+        return SOL_COAP_RESPONSE_CODE_BAD_REQUEST;
+    }
+
+    if (!obj_instance) {
+        SOL_WRN("Object instance was not provided."
+            " Can not complete the Bootstrap Write Resource operation");
+        return SOL_COAP_RESPONSE_CODE_BAD_REQUEST;
+    }
+
+    if (!payload.len) {
+        SOL_WRN("Payload to write on resource /%" PRIu16 "/%"
+            PRIu16 "/%" PRIu16 " is empty", obj_ctx->obj->id, obj_instance->id, resource_id);
+        return SOL_COAP_RESPONSE_CODE_BAD_REQUEST;
+    }
+
+    //Resource already exists?
+    if (resource_ctx) {
+        if (content_format == SOL_LWM2M_CONTENT_TYPE_TEXT ||
+            content_format == SOL_LWM2M_CONTENT_TYPE_OPAQUE) {
+            struct sol_lwm2m_resource res;
+            struct sol_blob *blob;
+
+            blob = sol_blob_new(&SOL_BLOB_TYPE_NO_FREE_DATA, NULL, payload.data, payload.len);
+            SOL_NULL_CHECK(blob, SOL_COAP_RESPONSE_CODE_BAD_REQUEST);
+
+            // TODO: what if I want to write INT, FLOAT, BOOLEAN?
```

Comment:
Right.

Section 6.3 from the spec was updated. It used to read _"The LWM2M Server MUST support all data formats. **The plain text and opaque formats MUST be supported by the LWM2M Client for singular Resources.** The LWM2M Client MUST support the TLV data format for Object Instance and multiple-instance Resource requests."_
Now it says _"The LWM2M Server MUST support all data formats. **The LWM2M Client MUST support the TLV data format for all requests.** In addition a LWM2M Client MAY choose to support other media type formats in the table below i.e., JSON, plain text and opaque."_

Anyway, I'll take note. Thanks!